### ISSUE TITLE
Automatic schema pull on write may fail if the keyspace is missing

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#3760]
If some node doesn't yet know about the keyspace or user defined types, but already receives writes for that keyspace, and user-defined types are present in table's definition which is referenced by the writes, then those writes will fail with the following error message:

```
storage_proxy - Failed to apply mutation from 10.195.15.220#28: std::_Nested_exception<schema_version_loading_failed> (Failed to load schema version 0b0c4cf0-181a-3527-83da-88b4561e3e7f): std::_Nested_exception<no_such_keyspace> (Can't find a keyspace ks_name): std::out_of_range (_Map_base::at)
```

It's because deserialization of schema definition requires instantiation of user-defined types, which requires table's keyspace to exist (see `cql_type_parser::parse()`). Keyspace and user types are not sent as part of the schema definition, so this will fail if the node doesn't yet know about keyspace and user types. We'd have to propagate keyspace and user types definition as well. Refs https://github.com/scylladb/scylla/issues/1207, https://github.com/scylladb/scylla/issues/2109

When schema is eventually propagated, the error will go away.

**Comment by tgrabiec** [comment]
The issue can happen even if there are no user-defined types, because the `db::cql_type_parser::parse()` looks up the keyspace unconditionally:

```
data_type db::cql_type_parser::parse(const sstring& keyspace, const sstring& str, lw_shared_ptr<user_types_metadata> user_types) {
    if (!user_types && service::get_storage_proxy().local_is_initialized()) {
        user_types = service::get_storage_proxy().local().get_db().local().find_keyspace(keyspace).metadata()->user_types();
    }

    auto raw = parse_raw(str);
    auto cql = raw->prepare_internal(keyspace, user_types);
    return cql.get_type();
}


data_type db::cql_type_parser::parse(const sstring& keyspace, const sstring& type) {
    return parse(keyspace, type, {});
}

```

**Comment by tgrabiec** [comment]
That's a regression since 7f64a6ec4bda96b53dc2f7fbc60b7ad5fd900c60

**Comment by yarongilor** [comment]
Issue is reproduced in 50GB longevity running version 3.1 rc6.
got thousands of the following error after running multiple-hard-reboot-nemesis:
```
< t:2019-09-09 20:59:28,095 f:sct_events.py   l:452  c:sdcm.sct_events      p:INFO  > 2019-09-09T20:58:47+00:00  ip-10-0-0-133 !WARNING | scylla: [shard 11] storage_proxy - Failed to apply mutation from 10.0.137.255#11: std::_Nested_exception<schema_version_loading_failed> (Failed to load schema version df909e7d-b645-339e-96c1-a1cf90c2b8ef): std::_Nested_exception<no_such_keyspace> (Can't find a keyspace keyspace_deflate): std::out_of_range (_Map_base::at)
```

**Comment by tgrabiec** [comment]
The general problem is that the schema definition of a table depends on the schema definition of the keyspace. Currently, this dependency is not carried together with the table schema definition.

Part of the fix could be to always make sure that the schema version is synced before unfreezing it. This would ensure that the local node's schema is at least as recent as the one referenced by the incoming request. So the current node would learn about keyspaces and UDTs, for example.

I think we need this regardless of problems with UDTs. If there is an incoming CL=1 read which reads from a newly created table we sould not fail that read because the replica hasn't learned about that table yet. That's not needed to unfreeze the schema definition but is needed for correctness. Requests carry a logical dependency on schema changes made prior to formation of the request on the coordinator node. We should carry that dependency to the replicas. Refs https://github.com/scylladb/scylla/issues/2109

There are still some problems though. This assumes that you can always unfreeze schema definition created at one point in time, at any later point in time, after arbitrary schema changes in between. This is not true if we allow UDTs or keyspaces to be deleted. If a UDT is deleted in between (together with columns using it), the request will fail to deserialize table's schema definition even if that request doesn't use the column which references the UDT.

Also, dropping a keyspace and recreating it in between will not be noticed, and we may assign an incorrect type upon schema definition deserialization. 

A solution for the two issues above would be to version the keyspace schema as well.

**Comment by tgrabiec** [comment]
Note that even if we revert 7f64a6e, so that we don't fail on table schema definition deserialization, we would still fail to serve the request because we need the keyspace to execute the request. We wouldn't get errors in replica logs, but those requests would still fail. User writes may succeed despite this if the CL is met.

**Comment by espindola** [comment]
> It's because deserialization of schema definition requires instantiation of user-defined types, which requires table's keyspace to exist (see `cql_type_parser::parse()`). Keyspace and user types are not sent as part of the schema definition, so this will fail if the node doesn't yet know about keyspace and user types.

Not sure if I am mixing two parts of the codebase, but include_keyspace_and_announce in migration_manager.cc always includes the keyspace when creating a mutation to avoid what I think was a similar issue found in cassandra: https://issues.apache.org/jira/browse/CASSANDRA-5631

Are we just missing the user defined types in there?

**Comment by tgrabiec** [comment]
On Wed, Sep 11, 2019 at 9:44 PM Rafael Ávila de Espíndola <
notifications@github.com> wrote:

> It's because deserialization of schema definition requires instantiation
> of user-defined types, which requires table's keyspace to exist (see
> cql_type_parser::parse()). Keyspace and user types are not sent as part
> of the schema definition, so this will fail if the node doesn't yet know
> about keyspace and user types.
>
> Not sure if I am mixing two parts of the codebase, but
> include_keyspace_and_announce in migration_manager.cc always includes the
> keyspace when creating a mutation to avoid what I think was a similar issue
> found in cassandra: https://issues.apache.org/jira/browse/CASSANDRA-5631
>
> Are we just missing the user defined types in there?
>

announcing is fine, the issue is with table schema definition requesting.

>

**Comment by kostja** [comment]
@tgrabiec I believe this is addressed by raft schema changes and can be closed. We only transfer schema using a migration manager with raft a as a whole chunk, so should transfer everything and apply in order. Do you agree?
If you have an idea how to construct a test case for the issue, please write it down here.

**Comment by tgrabiec** [comment]
@kostja It would be fixed if `migration_manager::get_schema_for_write()` would pull the latest schema before calling `get_schema_definition()`. It doesn't, local schema can lag behind the request's schema.

**Comment by mykaul** [comment]
@ptrsmrn - is this on track for 5.3?

**Comment by mykaul** [comment]
@kostja - is this something you plan to handle in 5.3?

**Comment by mykaul** [comment]
> @kostja - is this something you plan to handle in 5.3?

@kostja - let's make a decision on this (especially in the context of https://github.com/scylladb/scylladb/issues/13211 ?)

**Comment by tgrabiec** [comment]
The fix was partially reverted in 605e53e617812d9c26a24927167f23718cc8b396. We're still vulnerable during a time window when a restarting node is initializing group0 but other nodes already see it as UP and send requests to it.

**Comment by kostja** [comment]
Fixed for Raft mode in https://github.com/scylladb/scylladb/pull/14428 - there is still a window during upgrade when this can happen, which we don't plan to do anything about it.
